### PR TITLE
Fix empty ND-array construction

### PR DESCRIPTION
**Pull request by vrabaud**
This is currently allowed for a 2d matrix so let's allow it for an ND-array.
This fixes https://github.com/opencv/opencv/issues/27319

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [x] The PR is proposed to the proper branch
- [x] There is a reference to the original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

File: modules/core/test/test_mat.cpp [hunk 1]
```diff
@@ -1446,6 +1446,14 @@ TEST(Core_Mat, regression_9507)
     EXPECT_EQ(25u, m2.total());
 }
 
+TEST(Core_Mat, empty)
+{
+    // Should not crash.
+    uint8_t data[2] = {0, 1};
+    cv::Mat mat_nd(/*ndims=*/0, /*sizes=*/nullptr, CV_8UC1, /*data=*/data);
+    cv::Mat1b mat(0, 0, /*data=*/data, /*steps=*/1);
+}
```
Comment (vpisarev) [inline modules/core/test/test_mat.cpp:1455]: suggest to add some checks to the test:

```
TEST(Core_Mat, empty)
{
    // Should not crash.
    uint8_t data[2] = {0, 1};
    cv::Mat mat_nd(/*ndims=*/0, /*sizes=*/nullptr, CV_8UC1, /*data=*/data);
    cv::Mat1b mat(0, 0, /*data=*/data, /*steps=*/1);
    EXPECT_EQ(mat_nd.dims, 0);
    EXPECT_EQ(mat.dims, 2);
    EXPECT_LE(mat_nd.total(), 1u); // 0 for 4.x, 1 for 5.x
    EXPECT_EQ(mat.total(), 0u);
    EXPECT_TRUE(mat.empty());
    // mat_nd.empty() should return true for 4.x and false for 5.x
}
```

File: modules/core/src/matrix.cpp [hunk 2]
```diff
@@ -315,7 +315,7 @@ void finalizeHdr(Mat& m)
         m.rows = m.cols = -1;
     if(m.u)
         m.datastart = m.data = m.u->data;
-    if( m.data )
+    if( m.data && d > 0 )
```
Comment (vpisarev) [inline modules/core/src/matrix.cpp:318]: adding `d > 0` is fine.

I'd also suggest to modify the else branch:

```
else
        m.dataend = m.datalimit = 0;
```

to 

```
else
        m.dataend = m.datalimit = m.data;
```

that will make behaviour of both constructors of an empty matrix more consistent.

**Comment by opencv-alalek** [comment]
> 4.x

Does OpenCV 4.x support 0-D arrays (scalars)?

File: modules/core/src/matrix.cpp (hunk 2 above)
Comment (vrabaud) [reply modules/core/src/matrix.cpp:318]: Done

File: modules/core/test/test_mat.cpp (hunk 1 above)
Comment (vrabaud) [reply modules/core/test/test_mat.cpp:1466]: I fixed the test. @vpisarev , please decide with @opencv-alalek if it worth adding the test on total and empty.

**Review by vpisarev — APPROVED** [review]

**Review by asmorkalov — APPROVED** [review]
